packages/zoom-in/src/components/zoom.js

```diff
@@ -0,0 +1,259 @@
+import styled, { ThemeContext } from 'styled-components'
+import useEventListener from '../hooks/use-event-listener'
+
+const Container = styled.div`
+  position: relative;
+`
+
+const Original = styled.div`
+  visibility: hidden;
+  ${props =>
+    props.isZoomed
+      ? `position: absolute;
+		 left: 0;
+     top: 0;`
+      : `position: relative;`}
+`
+
+const Overlay = styled.div`
+  position: fixed;
+  top: 0;
+  right: 0;
+  bottom: 0;
+  left: 0;
+  opacity: 0;
+  transition: opacity ${props => props.transitionDuration}ms;
+  background: ${props => props.background};
+  ${props =>
+    props.show
+      ? `cursor: pointer;
+     cursor: zoom-out;`
+      : ''}
+`
+
+const Zoomable = styled.div`
+  cursor: pointer;
+  cursor: zoom-in;
+  ${props =>
+    props.isZoomed
+      ? `cursor: pointer;
+     cursor: zoom-out;`
+      : `position: absolute;
+		 left: 0;
+     top: 0;
+     `}
+`
+
+const Caption = styled.figcaption`
+  position: absolute;
+  top: 0;
+  left: 0;
+  opacity: 0;
+  visibility: hidden;
+  margin: 0;
+  color: ${props => props.color};
+  font-size: ${props => props.fontSize}px;
+  line-height: ${props => props.lineHeight}px;
+  letter-spacing: ${props => props.letterSpacing}px;
+  text-align: left;
+  ${props => (props.fontFamily ? `font-family: ${props.fontFamily};` : '')}
+`
+
+const Zoom = props => {
+  const themeContext = useContext(ThemeContext)
+  const { zoomOptions } = themeContext
+  const animate = animations[themeContext.caption.side] || animations.default
+  const [isZoomed, setZoom] = useState(false)
+  const overlayRef = useRef(null)
+  const originalRef = useRef(null)
+  const zoomedRef = useRef(null)
+  const containerRef = useRef(null)
+  const captionRef = useRef(null)
+
+  let isAnimating = false
+  const setAnimating = bool => {
+    isAnimating = bool
+  }
+
+  const getCaptionHeight = captionLength => {
+    const { fontSize, letterSpacing, width, lineHeight } = themeContext.caption
+    const height =
+      Math.ceil((captionLength * (fontSize + letterSpacing)) / width) *
+      lineHeight
+    return height
+  }
+
+  const handleOpenEnd = () => {
+    if (!isAnimating) return
+    if (!zoomedRef.current || !captionRef.current) return
+
+    setAnimating(false)
+    captionRef.current.style.opacity = 1
+    zoomedRef.current.removeEventListener('transitionend', handleOpenEnd)
+  }
+
+  let scrollTop = 0
+
+  const open = ({ target } = {}) => {
+    if (!target) return
+    if (isZoomed) return
+    if (!containerRef.current || !overlayRef.current || !zoomedRef.current)
+      return
+
+    const { overlay } = themeContext
+
+    scrollTop =
+      window.pageYOffset ||
+      document.documentElement.scrollTop ||
+      document.body.scrollTop ||
+      0
+
+    setAnimating(true)
+    containerRef.current.style.zIndex = overlay.zIndex
+    overlayRef.current.style.display = 'block'
+    window.requestAnimationFrame(() => {
+      overlayRef.current.style.opacity = overlay.opacity
+    })
```

Comment:
After a second thought, it is inappropriate to use `requestAnimationFrame` here.

According to [MDN web docs](https://developer.mozilla.org/zh-TW/docs/Web/API/Window.requestAnimationFrame), common use case is to draw animation step by step by repeatedly calling callback.  

Since the animation is executed by setting style directly, I don't think we need  `requestAnimationFrame` here. Will remove this in next updates.

Reply:
After a second thought, it seems inappropriate to use requestAnimationFrame here.
According to [MDN web docs](https://developer.mozilla.org/zh-TW/docs/Web/API/Window.requestAnimationFrame), common usage would be drawing animation step by step by repeatedly calling callback.

Since the animation is executed by setting style directly, I don't think we need `requestAnimationFrame` here. 

`requestAnimationFrame` provides delay time (1/60 sec) before executing callback.
But all we need to do is to seek a good timing to set `opacity` value of `overlay` node when the animation starts.

I'll remove `window.requestAnimationFrame` in next updates.